lib/active_hash/base.rb

```diff
@@ -136,23 +136,19 @@ def exists?(record)
 
       def insert(record)
         @records ||= []
-        record[:id] ||= next_id
+        @max_id ||= 0
+        if record[:id] && record[:id].is_a?(Numeric)
+          @max_id = [@max_id, record[:id]].max
+        else
+          record[:id] ||= (@max_id = @max_id.succ)
+        end
         validate_unique_id(record) if dirty
```

Comment:
It is possible to create an object with a negative id. In this case the `max_id` doesn't get updated correctly by the suggested edit. The previous behaviour would correctly identify the maximum id in this case.

There was also an issue with the previous code that `.succ` exists only for `Integer` and not for `Numeric` so it would raise a `NoMethodError Exception: undefined method 'succ' for 1.0:Float` if an object with a `Float` id happens to have the maximum `id`. This issue persists in the suggested edit.

I think it might be better to keep the id logic in a separate function to not clutter the insert logic. So I suggest the following (untested) code:

```ruby
      def insert(record)
        @records ||= []
        set_id(record)
        validate_unique_id(record) if dirty
        ...
      end

      begin set_id(record)
        # sets record[:id] to @max_id+1 if it doesn't exist
        if record[:id] && record[:id].is_a?(Numeric)
          @max_id = [@max_id || -Float::INFINITY, record[:id].ceil].max
        else
          record[:id] ||= (@max_id = @max_id.to_i.succ)
        end
      end
```
@braktar let me know if you like the suggested edit. Either I can include it in my PR or commit to your branch.

Reply:
I just prefer to be consistent with the current tests and initialize @max_id to 0

Reply:
In that case can you add a test that creates an object with negative ID and then another one without id, to check the id of the last one to be 1.

You still take care of the Float id problem which can be handled with `.ceil` .

Reply:
Oh I just saw the new commit, you already did the test :facepalm: